src/darwin/Framework/CHIP/MTRCommissioningOperation.mm

```diff
@@ -0,0 +1,490 @@
+        [self _earlyFailCommissioning:CHIP_ERROR_BUSY];
+        return;
+    }
+
+    MTR_LOG("%@ starting commissioning with controller %@", self, _controller);
+
+    if (!_isInternallyCreated) {
+        [[MTRMetricsCollector sharedInstance] resetMetrics];
+
+        // Track overall commissioning
+        MATTER_LOG_METRIC_BEGIN(kMetricDeviceCommissioning);
+        emitMetricForSetupPayload(_setupPayload);
+    }
+
+    CHIP_ERROR err = [concreteController startCommissioning:self withCommissioningID:_commissioningID];
+    if (err != CHIP_NO_ERROR) {
+        MTR_LOG_ERROR("%@ failed to start commissioning with controller %@", self, _controller);
+        [self _dispatchCommissioningCHIPError:err];
+    }
+}
+
+- (BOOL)stop
+{
+    MTRDeviceController_Concrete * strongController = _controller;
+    if (!strongController) {
+        // Nothing to do; controller is gone, so we are stopped no matter what.
+        return NO;
+    }
+
+    return [strongController stopCommissioning:self forCommissioningID:_commissioningID];
+}
+
+- (void)_earlyFailCommissioning:(CHIP_ERROR)error
+{
+    // This handles failures before we have cleared the way for doing a commissioning at all, and in
+    // particular we might have an existing commissioning when we land here.  Don't mess with the
+    // "normal" metrics here, because we might stomp on the metrics for that ongoing commissioning.
+    auto * emptyMetrics = [[MTRMetrics alloc] initWithCapacity:0];
+    [self _dispatchCommissioningError:[MTRError errorForCHIPErrorCode:error] withMetrics:emptyMetrics];
+}
+
+- (void)_dispatchCommissioningCHIPError:(CHIP_ERROR)error
+{
+    MATTER_LOG_METRIC_END(kMetricDeviceCommissioning, error);
+    MTRMetrics * metrics = [[MTRMetricsCollector sharedInstance] metricSnapshotForCommissioning:YES];
+    [self _dispatchCommissioningError:[MTRError errorForCHIPErrorCode:error] withMetrics:metrics];
+}
+
+- (void)_dispatchCommissioningError:(NSError *)error
+{
+    MATTER_LOG_METRIC_END(kMetricDeviceCommissioning, [MTRError errorToCHIPErrorCode:error]);
+    MTRMetrics * metrics = [[MTRMetricsCollector sharedInstance] metricSnapshotForCommissioning:YES];
+    [self _dispatchCommissioningError:error withMetrics:metrics];
+}
+
+- (void)_dispatchCommissioningError:(NSError *)error withMetrics:(MTRMetrics *)metrics
+{
+    [self _dispatchCommissioningError:error forCommissioningID:_commissioningID withMetrics:metrics];
+}
+
+- (void)_dispatchCommissioningError:(NSError *)error forCommissioningID:(NSNumber *)commissioningID withMetrics:(MTRMetrics *)metrics
+{
+    MTRDeviceController_Concrete * strongController = _controller;
+
+    MTR_LOG("%@ Device commissioning failed with controller %@ metrics %@", self, strongController, metrics);
+
+    id<MTRCommissioningDelegate> strongDelegate = _delegate;
+    id<MTRCommissioningDelegate_Internal> strongInternalDelegate = [self _internalDelegate];
```

Comment:
Why are you distinguishing `strongDelegate` and `strongInternalDelegate` here? `_internalDelegate` just performs an unchecked downcast, so they are always going to be the same pointer.